### DISCUSSION TITLE
help creating pattern to remove linkid from amazon

### DISCUSSION THREAD

**Discussion by pokegamer5547** [TrianguloY/URLCheck#400] [Q&A] [answered]
Hello um i need some help with regex patterns. I'm trying to remove linkid param from amazon links but i keep failing. I tried using this regex made by chatgpt but it still keeps question mark.

"remove linkId from Amazon URL": {
  "regex": "^(https?:\\/\\/www\\.amazon\\.in\\/[^?]*)(\\?.*?)?(?:&?linkId=[^&]*)(?:&)?(.*)$",
  "replacement": "$1$2$3",
  "automatic": "true"
}

https://www.amazon.in/shop/MrSujano?linkId=69900caff6b8ede42352973921a6ee43&language=en_US

https://www.amazon.in/gp/product/B08D45C1YF?linkId=3fb12d772aedc6efe071703f0b838a7a

**Comment by PabloOQ** [comment, marked as answer]
I think the URL Cleaner module is better suited for this task, or at least makes more sense to me.

Just edit the file like the following. Coincidentally, amazon is the first that shows in the built in catalog, which is where I get the `urlPattern` and `exceptions` fields
```json
{
  "customRules": {
    YOUR RULES HERE
  },
  "providers": {
    "amazon": {
... rest of the file ...
```
And in YOUR RULES HERE paste the following:
```json
    "amazon": {
      "urlPattern": "^https?:\\/\\/(?:[a-z0-9-]+\\.)*?amazon(?:\\.[a-z]{2,}){1,}",
      "rules": [
        "linkId"
        ],
        "exceptions": [
          "^https?:\\/\\/(?:[a-z0-9-]+\\.)*?amazon(?:\\.[a-z]{2,}){1,}\\/gp\\/.*?(?:redirector.html|cart\\/ajax-update.html|video\\/api\\/)",
          "^https?:\\/\\/(?:[a-z0-9-]+\\.)*?amazon(?:\\.[a-z]{2,}){1,}\\/(?:hz\\/reviews-render\\/ajax\\/|message-us\\?|s\\?)"
        ]
      }
```

In fact, here is my custom entry if you want it. Note that try to strip every parameter I can, and while I do testing, maybe one of these parameters is necessary for something and I just remove it for no reason.
```json
  "amazon": {
      "urlPattern": "^https?:\\\/\\\/(?:[a-z0-9-]+\\.)*?amazon(?:\\.[a-z]{2,}){1,}",
      "rules": [
        "psc",
        "spc",
        "sp_csd",
        "linkId",
        "sbo",
        "xpid",
        "ufe",
        "spLa",
        "maas",
        "btn_ref",
        "social_share",
        "skipTwisterOG",
        "starsLeft",
        "visitId"
      ],
      "exceptions": [
        "^https?:\\\/\\\/(?:[a-z0-9-]+\\.)*?amazon(?:\\.[a-z]{2,}){1,}\\\/gp\\\/.*?(?:redirector.html|cart\\\/ajax-update.html|video\\\/api\\\/)",
        "^https?:\\\/\\\/(?:[a-z0-9-]+\\.)*?amazon(?:\\.[a-z]{2,}){1,}\\\/(?:hz\\\/reviews-render\\\/ajax\\\/|message-us\\?|s\\?)"
      ]
    }
```

**Reply by pokegamer5547** [reply]
> I think the URL Cleaner module is better suited for this task, or at least makes more sense to me.
> 
> Just edit the file like the following. Coincidentally, amazon is the first that shows in the built in catalog, which is where I get the `urlPattern` and `exceptions` fields
> 
> ```json
> {
>   "customRules": {
>     YOUR RULES HERE
>   },
>   "providers": {
>     "amazon": {
> ... rest of the file ...
> ```
> 
> And in YOUR RULES HERE paste the following:
> 
> ```json
>     "amazon": {
>       "urlPattern": "^https?:\\/\\/(?:[a-z0-9-]+\\.)*?amazon(?:\\.[a-z]{2,}){1,}",
>       "rules": [
>         "linkId"
>         ],
>         "exceptions": [
>           "^https?:\\/\\/(?:[a-z0-9-]+\\.)*?amazon(?:\\.[a-z]{2,}){1,}\\/gp\\/.*?(?:redirector.html|cart\\/ajax-update.html|video\\/api\\/)",
>           "^https?:\\/\\/(?:[a-z0-9-]+\\.)*?amazon(?:\\.[a-z]{2,}){1,}\\/(?:hz\\/reviews-render\\/ajax\\/|message-us\\?|s\\?)"
>         ]
>       }
> ```
> 
> In fact, here is my custom entry if you want it. Note that try to strip every parameter I can, and while I do testing, maybe one of these parameters is necessary for something and I just remove it for no reason.
> 
> ```json
>   "amazon": {
>       "urlPattern": "^https?:\\\/\\\/(?:[a-z0-9-]+\\.)*?amazon(?:\\.[a-z]{2,}){1,}",
>       "rules": [
>         "psc",
>         "spc",
>         "sp_csd",
>         "linkId",
>         "sbo",
>         "xpid",
>         "ufe",
>         "spLa",
>         "maas",
>         "btn_ref",
>         "social_share",
>         "skipTwisterOG",
>         "starsLeft",
>         "visitId"
>       ],
>       "exceptions": [
>         "^https?:\\\/\\\/(?:[a-z0-9-]+\\.)*?amazon(?:\\.[a-z]{2,}){1,}\\\/gp\\\/.*?(?:redirector.html|cart\\\/ajax-update.html|video\\\/api\\\/)",
>         "^https?:\\\/\\\/(?:[a-z0-9-]+\\.)*?amazon(?:\\.[a-z]{2,}){1,}\\\/(?:hz\\\/reviews-render\\\/ajax\\\/|message-us\\?|s\\?)"
>       ]
>     }
> ```

Thankyou so much. you're right urlcleaner is better suited. But my problem with it is that it'll get overwritten if it updates the clearurl json :(

**Reply by PabloOQ** [reply]
It won't! (at least not anymore, if I remember correctly it used to work like that once) It ONLY overwrites what is inside the parent object `"providers"`, that's why I added a new `"customRules"` parent object, here is where you must put your rules. I'm not sure if you can have as many custom parent objects as you want, but at the very least you can have one additional parent object.

```json
{
  "customRules": {
    this will NOT get overwritten
  },
  "providers": {
    this will get overwritten
  }
}
```

Note that when checking, for "nasty" urls we check all the rules, so adding a new one with the exact same `"urlPattern"` doesn't mean it will skip the "clone", so you can freely copy from the `"providers"` as a template.

**Reply by pokegamer5547** [reply]
> It won't! (at least not anymore, if I remember correctly it worked like that once) It ONLY overwrites what is inside the parent object `"providers"`, that's why I added a new `"customRules"` parent object, here is where you must put your rules. I'm not sure if you can have as many custom parent objects as you want, but at the very least you can have one additional parent object.
> 
> ```json
> {
>   "customRules": {
>     this will NOT get overwritten
>   },
>   "providers": {
>     this will get overwritten
>   }
> }
> ```
> 
> Note that when checking, for "nasty" urls we check all the rules, so adding a new one with the exact same "urlPattern" doesn't mean it will skip the "clone", so you can freely copy from the `"providers"` as a template.


Ah i see, i understand now. So long as i use seperate parent it wont be affected.
Thankyou so much :)

**Reply by TrianguloY** [reply]
> It won't! (at least not anymore, if I remember correctly it used to work like that once)

Confirmed! :)
